File: csharp/binary-search/BinarySearch.cs
using System;

public static class BinarySearch
{
    public static int Find(int[] input, int value) =>
        input.Length == 0
            ? -1
            : SearchIndex(input, value, 0, input.Length - 1);

    private static int SearchIndex(int[] input, int value, int minIndex, int maxIndex)
    {
        var halveIndex = (minIndex + maxIndex) / 2;

        return input[halveIndex] == value
            ? halveIndex
            : halveIndex <= 0 || halveIndex >= maxIndex || minIndex == maxIndex
            ? -1
            : input[halveIndex] > value
            ? SearchIndex(input, value, minIndex, halveIndex - 1)
            : SearchIndex(input, value, halveIndex + 1, maxIndex);
    }
}

File: fsharp/binary-search/BinarySearch.fs
module BinarySearch

let rec private binarySearch index value =
    function
    | [||] ->
        None
    | input ->   
        let middle = input.Length / 2
        let middleValue = input.[middle]

        match value with
        | value when value < middleValue ->
            input.[0 .. middle - 1] |> binarySearch index value
        | value when value > middleValue ->
            input.[middle + 1 ..] |> binarySearch (index + middle + 1) value 
        | _ ->
            Some (index + middle)

let find input value =
    match (input |> Array.sort) <> input with
    | true -> failwith "The input must be an ordered lists"
    | false -> input |> binarySearch 0 value